src/reader.cpp

```diff
@@ -1173,9 +1183,24 @@ static std::function<autosuggestion_result_t(void)> get_autosuggestion_performer
         if (!cursor_at_end && iswspace(last_char)) return nothing;
 
         // On the other hand, if the line ends with a quote, don't go dumping stuff after the quote.
-        if (wcschr(L"'\"", last_char) && cursor_at_end) return nothing;
+        if ((last_char == L'"' || last_char == L'\'') && cursor_at_end) return nothing;
+
+        // Try normal completions last
+        // But not unless the head has been fully typed out and exists/is valid.
+        const wchar_t *first_space = wcschr(first_char, L' ');
+        if (first_space == nullptr) {
+            return nothing; //the head has not been typed out yet
+        }
+        wcstring head(first_char, first_space);
+
+        // Ideally, we should use function_exists and not function_exists_no_autoload here,
```

Comment:
Why? The "no_autoload" version can still tell you if the function exists (or rather is _supposed_ to exist - i.e. a file that should contain a function of that name does).

Reply:
I didn't realize that. Awesome.